### PR TITLE
basic SVGD implementation

### PR DESCRIPTION
**Pull request by martinjankowiak**
pair-coded with @fritzo.

turned out to be pretty simple thanks to existing `AutoDelta` machinery.

[paper reference](https://arxiv.org/abs/1608.04471)

### CODE REVIEW

File: pyro/infer/svgd.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    References:
+    [1] "Stein Variational Gradient Descent: A General Purpose Bayesian Inference Algorithm,"
+        Qiang Liu, Dilin Wang
+    """
+    def __init__(self, model, kernel, num_particles, max_plate_nesting):
+        self.model = vectorize(model, num_particles, max_plate_nesting)
+        # TODO: fix circular import hack
+        from pyro.contrib.autoguide import AutoDelta
+        self.guide = AutoDelta(model)
```
Comment (fritzo) [inline pyro/infer/svgd.py]: `model` -> `self.model`

File: pyro/contrib/autoguide/__init__.py [hunk 2]
```diff
@@ -306,7 +306,11 @@ def __call__(self, *args, **kwargs):
                 value = pyro.param("{}_{}".format(self.prefix, name),
                                    site["value"].detach(),
                                    constraint=site["fn"].support)
-                result[name] = pyro.sample(name, dist.Delta(value, event_dim=site["fn"].event_dim))
+                transform = biject_to(site["fn"].support)
```
Comment (martinjankowiak) [inline pyro/contrib/autoguide/__init__.py]: @fritzo had to add this stuff. besides letting this log det jacobian logic be controlled by a flag (`include=True/False`) i could instead introduce a subclass:
```class SVGDGuide(AutoDelta):....```

File: pyro/infer/svgd.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+
+def vectorize(fn, num_particles, max_plate_nesting):
+    def _fn(*args, **kwargs):
+        with pyro.plate("num_particles_vectorized", num_particles, dim=-max_plate_nesting - 1):
+            return fn(*args, **kwargs)
+    return _fn
+
+
+class SVGDKernel(object):
```
Comment (martinjankowiak) [inline pyro/infer/svgd.py]: could also call this a `SteinKernel` or the like (as in KSD = kernelized stein discrepancy)

File: pyro/infer/svgd.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            with ExitStack() as stack:
+                for frame in site["cond_indep_stack"]:
+                    if frame.vectorized:
+                        stack.enter_context(plates[frame.name])
+                value = pyro.param("{}_{}".format(self.prefix, name),
+                                   site["value"].detach(),
+                                   constraint=site["fn"].support)
+                # The following lines are where this method differs from the AutoDelta version.
+                # For SVGD (but not MAP) we need the log det jacobian terms for correctness.
+                transform = biject_to(site["fn"].support)
```
Comment (martinjankowiak) [inline pyro/infer/svgd.py]: @fritzo please let me know if you have any suggestions for how to clean this part up. this machinery is a bit opaque to me

File: pyro/infer/svgd.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    Example usage:
+
+    >>> from pyro.infer import SVGD, RBFSteinKernel
+    >>> from pyro.optim import Adam
+
+    >>> kernel = RBFSteinKernel()
+    >>> adam = Adam({"lr": 0.1})
+    >>> svgd = SVGD(model, kernel, adam, num_particles=50, max_plate_nesting=0)  # doctest: +SKIP
+
+    >>> def gradient_callback(squared_gradients):
```
Comment (martinjankowiak) [inline pyro/infer/svgd.py]: added a gradient callback

File: pyro/infer/svgd.py (hunk 4 above)
Comment (martinjankowiak) [reply pyro/infer/svgd.py]: @fritzo pending any suggestions on this bit, i'd like to get this PR merged. this is because i think i'm going to have to refactor things a bit anyway to support a cleaner/more performant/more general version in #1991. that will also help me tackle any lingering shape issues (e.g. scalar random variables). lmk if you have any questions

Comment (fritzo) [reply pyro/infer/svgd.py]: TODO: please try to factor this code and share some code with pyro/infer/autoguide/guides.py. Maybe add a helper in autoguides/guide.py. This is very tricky code and just had a bug fixed a few weeks ago. If we copy and paste, the two bits will diverge.

File: pyro/infer/svgd.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+                transform = biject_to(site["fn"].support)
+                unconstrained_value = pyro.param("{}_{}".format(self.prefix, name)).unconstrained()
+                log_density = transform.inv.log_abs_det_jacobian(value, unconstrained_value)
+                log_density = sum_rightmost(log_density, log_density.dim() - value.dim() + site["fn"].event_dim)
+                result[name] = pyro.sample(name, Delta(value, log_density=log_density,
+                                           event_dim=site["fn"].event_dim))
+        return result
+
+
+class SteinKernel(object):
```
Comment (fritzo) [inline pyro/infer/svgd.py]: TODO: make this into an abstract base class with an `@abstractmethod` `.kernel_and_grads()`.

**Comment by martinjankowiak** [comment]
@fritzo thanks for the review and TODOs! this will indeed benefit from further clean-up and refactoring
